### PR TITLE
Followers feature

### PR DESCRIPTION
**Pull request by rmonteseba**
New followers capability with domain restrictions, comments-user relations, and  timestamps visualization

### CODE REVIEW

File: app/controllers/comments_controller.rb [hunk 1]
```diff
@@ -26,6 +35,6 @@ def destroy
   private
 
   def comment_params
-    params.require(:comment).permit(:commenter, :body, :status)
+    params.require(:comment).permit(:body, :status)
```
Comment (juanhurtado10) [inline app/controllers/comments_controller.rb]: @JsMontesR Could you please explain what this line does?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if @user.update(user_params)
+      redirect_to root_path, notice: 'Your info have been updated!'
+    else
+      render :edit
+    end
+  end
+
+  # Makes the logged user to follow the param user
+  def add_follower
+    @user = User.find(params[:user_id])
```
Comment (juanhurtado10) [inline app/controllers/users_controller.rb]: @JsMontesR what if the user is nil 🤔?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# The User model
+class User < ApplicationRecord
+  attr_accessor :old_password
+
+  has_many :articles, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_and_belongs_to_many :followers,
```
Comment (juanhurtado10) [inline app/models/user.rb]: @JsMontesR Could you please explain this line, and its attributes?

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,2 +1,15 @@
 class ApplicationController < ActionController::Base
+  helper_method :current_user
+
+  def current_user
+    if session[:user_id]
+      begin
+        @current_user ||= User.find(session[:user_id])
+      rescue ActiveRecord::RecordNotFound
```
Comment (juanhurtado10) [inline app/controllers/application_controller.rb]: @JsMontesR I like to see this. Why is this required here?

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -1,19 +1,28 @@
...
 
-  http_basic_authenticate_with name: "sebas", password: "secret", only: :destroy
+  include PrivateGuard
+
+  before_action :require_user
 
   def create
     @article = Article.find(params[:article_id])
-    @comment = @article.comments.create(comment_params)
+    @comment = Comment.create(comment_params)
```
Comment (juanhurtado10) [inline app/controllers/comments_controller.rb]: @JsMontesR What if the creation fails for any reason?

File: app/models/comment.rb [hunk 6]
```diff
@@ -1,4 +1,9 @@
 class Comment < ApplicationRecord
   include Visible
+  include DateAccessor
   belongs_to :article
+  belongs_to :user
+
+  validates_presence_of :body
```
Comment (juanhurtado10) [inline app/models/comment.rb]: @JsMontesR What this line does?

File: app/models/concerns/visible.rb [hunk 7]
```diff
@@ -1,7 +1,10 @@
 module Visible
   extend ActiveSupport::Concern
+  PUBLIC = 'public'.freeze
```
Comment (juanhurtado10) [inline app/models/concerns/visible.rb]: @JsMontesR Why .`freeze` is required here?

File: app/views/articles/index.html.erb [hunk 8]
```diff
@@ -1,30 +1,12 @@
...
-    <% unless article.archived? %>
-      <li>
-        <%= link_to article.title, article %>
-      </li>
-    <% end %>
-  <% end %>
-</ul>
-
-<%= link_to "New Article", new_article_path %>
+<h2>Our blog has <%= Article.count_without_status('archived') %> active articles and counting!</h2>
```
Comment (juanhurtado10) [inline app/views/articles/index.html.erb]: @JsMontesR Is it ok having logic in the views? What do you think?

File: app/views/comments/edit.html.erb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<h1>Edit comment</h1>
+<%= form_with model: [@article, @comment] do |form| %>
```
Comment (juanhurtado10) [inline app/views/comments/edit.html.erb]: @JsMontesR Please explain what this line does?

File: app/views/sessions/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  <div class="field">
+    <%= label_tag :password %><br>
+    <%= password_field_tag :password %>
+  </div>
+  <br>
+  <div class="actions">
+    <%= submit_tag "Login" %>
+  </div>
+
+<% end %>
```
Comment (juanhurtado10) [inline app/views/sessions/new.html.erb]: @JsMontesR As a good practice tip: keep a blank space at the end of the files

File: app/views/shared/_status.html.erb [hunk 11]
```diff
@@ -1,9 +1,6 @@
 <p>
-  <%= form.label "visibility:" %><br>
+  <%= form.label "Visibility:" %><br>
   <%= form.select :status,
-                  [["Public", "public"], ["Private", "private"], ["Archived", "archived"]],
-                  selected: "public" %>
-  <%= element.errors.full_messages_for(:status).each do |message| %>
-  <div><%= message %></div>
-<% end %>
+                  [['Public', 'public'], ['Personal', 'personal'], ['Archived', 'archived']],
```
Comment (juanhurtado10) [inline app/views/shared/_status.html.erb]: @JsMontesR What if we move this trivial logic into helper methods? Does it make sense?

File: config/routes.rb [hunk 12]
```diff
@@ -1,6 +1,15 @@
 Rails.application.routes.draw do
-  root ("articles#index") #get "/articles", to: "articles#index"
+  root ("home#index")
   resources :articles do
     resources :comments
   end
+  resources :users, except: %i[destroy] do
```
Comment (juanhurtado10) [inline config/routes.rb]: @JsMontesR What does the resources helper method?

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (rmonteseba) [reply app/controllers/comments_controller.rb]: Sure, the purpose of this line is filter the incoming petition, in this case it is requiring a :comment as a principal incoming key with the subsequent required fields :body and :status. It defines something like..

{ comment: { 
boby:  foo1, 
status: foo2
} }

In a few words there is being defined strong parameters, whatever fields witihin an incoming petition that doesnt match that structure will be skipped, if some of the args within the basic strcuture is not specified within the incoming petition it raises an error.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rmonteseba) [reply app/controllers/users_controller.rb]: Fatal!, if user is not found that code-line will raise an exception (ActiveRecord::RecordNotFound), is necessary a validation for handling that error in this context, I think personally for making explicit the 500 error and show the error page or something

File: app/models/user.rb (hunk 3 above)
Comment (rmonteseba) [reply app/models/user.rb]: "has_many" as its name suggest, indicates that the current model in this case the "user" model, has many references to another model mapped by a relation-name, Rails handle those relations names following a convention, in a few words, an user model instance can reference to all its articles through the articles relation (user.articles), wich following the convention, the name "articles" references to the "article" model in singular. The "dependent:" atribute basically specify the "on_delete" action on the DB side, in this case when a user is dropped, all its articles and comments are also dropped.

The " has_and_belongs_to_many" keyword is used for defining many to many relations on the same basis previously explained, with the diference that now the current model also belongs to other model (in other words, other model has many references to some of the current model instances), and in this case down bellow the referenced lines, it was necesary to specify some additional parameters for defining the relation name and some database-relation aditional stuff (ex: target model, table name and FKs).

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (rmonteseba) [reply app/controllers/application_controller.rb]: The if condition is necesary for find out if the user has a current sesion, that validation within the body was necessary because when the "current_user" is required it can exist in the session context but no always in the DB context (for example if the user was deleted while its session was active).

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rmonteseba) [reply app/controllers/users_controller.rb]: Oh, okay, ActiveRecord::RecordNotFound o whatever exception would be, depending of the context.

File: app/controllers/comments_controller.rb (hunk 5 above)
Comment (rmonteseba) [reply app/controllers/comments_controller.rb]: Same that the last case, fatal!, it raises the respective exception (a validation error preferably) when the "save!" method is called down bellow.

File: app/models/comment.rb (hunk 6 above)
Comment (rmonteseba) [reply app/models/comment.rb]: belongs_to refers that the current model belongs to another model, an user model in this case, and the "validates_presence_of", verifies that the incoming "body" of the model thats being saved is not blank ("") or nil.

File: app/models/concerns/visible.rb (hunk 7 above)
Comment (rmonteseba) [reply app/models/concerns/visible.rb]: ".freeze" freezes the target object, that because I don't want it to mutate along time, I want it constant.
That was a rubo-cop recomendation.

File: app/views/articles/index.html.erb (hunk 8 above)
Comment (rmonteseba) [reply app/views/articles/index.html.erb]: That's horrible!, I'd prefer to create a global helper (if it is possible) something like "count_archived_articles" and handle its logic from outside the view.

File: app/views/comments/edit.html.erb (hunk 9 above)
Comment (rmonteseba) [reply app/views/comments/edit.html.erb]: This line creates a form based on the comment model and it procedence relation (comming from articles), so aside from also having the article reference within the form, when submiting the form the request will fetch a URL based on that procedence (and also dependence) relation, this to something like /articles/foo_id/comments.... instead of only begin by /comments/... as itself.

File: app/views/sessions/new.html.erb (hunk 10 above)
Comment (rmonteseba) [reply app/views/sessions/new.html.erb]: I'll keep in mind! 👌🏽 , btw, why?

File: app/views/shared/_status.html.erb (hunk 11 above)
Comment (rmonteseba) [reply app/views/shared/_status.html.erb]: That make sense for me, maybe with the helper strategy we can access the concern statuses and render them here. That would be better I think. 🤔

File: config/routes.rb (hunk 12 above)
Comment (rmonteseba) [reply config/routes.rb]: The resources helper method actually enables all the CRUD bases endpoints for an specific model, it allows some aditional configuration for select only some kind of operations (for this example, all excepts the "destroy" capability).

**Comment by rmonteseba** [comment]
I've not resolved the conflicts yet until the user-layout #2 feature be merged.

File: app/controllers/comments_controller.rb (hunk 1 above)
Comment (juanhurtado10) [reply app/controllers/comments_controller.rb]: @JsMontesR Good ✅

File: app/models/user.rb (hunk 3 above)
Comment (juanhurtado10) [reply app/models/user.rb]: @JsMontesR Excellent ✅

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (juanhurtado10) [reply app/controllers/application_controller.rb]: @JsMontesR Good ✅

File: app/views/articles/index.html.erb (hunk 8 above)
Comment (juanhurtado10) [reply app/views/articles/index.html.erb]: @JsMontesR that's correct, please fix it in your code!

File: app/views/comments/edit.html.erb (hunk 9 above)
Comment (juanhurtado10) [reply app/views/comments/edit.html.erb]: @JsMontesR Outstanding ✅

File: app/views/shared/_status.html.erb (hunk 11 above)
Comment (juanhurtado10) [reply app/views/shared/_status.html.erb]: @JsMontesR I think it would be a better approach!

File: config/routes.rb (hunk 12 above)
Comment (juanhurtado10) [reply config/routes.rb]: @JsMontesR Really good ✅.

File: app/controllers/comments_controller.rb (hunk 5 above)
Comment (juanhurtado10) [reply app/controllers/comments_controller.rb]: @JsMontesR Good ✅, please try to fix it!

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (juanhurtado10) [reply app/controllers/users_controller.rb]: @JsMontesR You right, please try to solve that use case in your code!!

**Comment by juanhurtado10** [comment]
@JsMontesR please do a rebase to remove these conflicts!

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (rmonteseba) [reply app/controllers/users_controller.rb]: ✅

File: app/controllers/comments_controller.rb (hunk 5 above)
Comment (rmonteseba) [reply app/controllers/comments_controller.rb]: ✅

File: app/views/articles/index.html.erb (hunk 8 above)
Comment (rmonteseba) [reply app/views/articles/index.html.erb]: ✅

File: app/views/shared/_status.html.erb (hunk 11 above)
Comment (rmonteseba) [reply app/views/shared/_status.html.erb]: ✅ Done!

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -1,2 +1,27 @@
 class ApplicationController < ActionController::Base
+  helper_method :current_user
+  helper_method :count_unarchived_articles
+  helper_method :valid_statuses
+
+  def current_user
+    if session[:user_id]
```
Comment (juanhurtado10) [inline app/controllers/application_controller.rb]: @JsMontesR A quick return here will make your code easy to understand.
```
def current_user
  @current_user = nil
    return @current_user unless session[:user_id]
    begin
        @current_user ||= User.find(session[:user_id])
    rescue ActiveRecord::RecordNotFound
        @current_user
    end
end
```

File: app/controllers/articles_controller.rb [hunk 14]
```diff
@@ -1,54 +1,55 @@
...
   end
 
   def update
-    @article = Article.find(params[:id])
-
-    if @article.update(article_params)
-      redirect_to @article
-    else
-      render :edit
+    if (@article = find(current_user.articles, params[:id])).is_a?(Article) # current_user.articles.find(params[:id])
```
Comment (juanhurtado10) [inline app/controllers/articles_controller.rb]: @JsMontesR too much logic to be placed in a Controller, moving logic into a Class is a common pattern ... just as a tip.

File: app/controllers/articles_controller.rb [hunk 15]
```diff
@@ -1,54 +1,55 @@
...
+    if @article
       redirect_to @article
     else
       render :new
     end
   end
 
   def edit
-    @article = Article.find(params[:id])
+    @article = find(current_user.articles, params[:id]) # current_user.articles.find(params[:id])
```
Comment (juanhurtado10) [inline app/controllers/articles_controller.rb]: @JsMontesR the **find** method is not declarative enough

File: app/controllers/comments_controller.rb [hunk 16]
```diff
@@ -1,31 +1,52 @@
...
+  include PrivateGuard
+  include SecureTransaction
+
+  before_action :require_user
 
   def create
-    @article = Article.find(params[:article_id])
-    @comment = @article.comments.create(comment_params)
-    redirect_to article_path(@article)
+    if (@article = find(Article, params[:article_id])).is_a?(Article)
```
Comment (juanhurtado10) [inline app/controllers/comments_controller.rb]: @JsMontesR again, too much logic that should not be placed in a controller, as a tip: "Controllers should be just in charge of invoking "operations" and rendering "templates"

File: app/controllers/concerns/secure_transaction.rb [hunk 17]
```diff
@@ -0,0 +1,28 @@
+# Module that allows some transactions over DB resources on a safer way
+module SecureTransaction
```
Comment (juanhurtado10) [inline app/controllers/concerns/secure_transaction.rb]: @JsMontesR This is a good approach 👏🏻

File: db/seeds.rb [hunk 18]
```diff
@@ -5,3 +5,9 @@
 #
 #   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
 #   Character.create(name: 'Luke', movie: movies.first)
+
+5.times do |i|
```
Comment (juanhurtado10) [inline db/seeds.rb]: @JsMontesR Awesome, Really happy to see seeds! 👏🏻
